src/details/ArborX_DetailsDistributedTreeImpl.hpp

```diff
@@ -671,6 +675,128 @@ DistributedTreeImpl<DeviceType>::queryDispatch(
   Kokkos::Profiling::popRegion();
 }
 
+template <typename Query>
+struct QueriesWithIndices
+{
+  Query query;
+  int query_id;
+  int primitive_index;
+};
+
+template <typename DeviceType>
+template <typename DistributedTree, typename ExecutionSpace,
+          typename Predicates, typename OutputView, typename OffsetView,
+          typename Callback>
+std::enable_if_t<Kokkos::is_view<OutputView>{} && Kokkos::is_view<OffsetView>{}>
+DistributedTreeImpl<DeviceType>::queryDispatch(
+    NearestPredicateTag, DistributedTree const &tree,
+    ExecutionSpace const &space, Predicates const &queries,
+    Callback const &callback, OutputView &out, OffsetView &offset)
+{
+  using Indices = Kokkos::View<int *, ExecutionSpace>;
+  Indices indices("ArborX::DistributedTree::query::nearest::indices", 0);
+  Kokkos::View<int *, ExecutionSpace> ranks(
+      "ArborX::DistributedTree::query::nearest::ranks", 0);
+
+  // First find the ranks and indices for the nearest queries.
+  queryDispatchImpl(NearestPredicateTag{}, tree, space, queries, indices,
+                    offset, ranks);
```

Comment:
Please expand the comments on the strategy here, that we have another round of communication after we reduced the final set of nearest neighbors, that we can't just call the callbacks the first time around, etc.